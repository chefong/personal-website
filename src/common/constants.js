import hamburgerMenuAnimation from '~/assets/menu-animation.json';

export const gaEvents = {
    PAGE_RENDERED: 'page_rendered',
    ABOUT_NAV_CLICKED: 'about_nav_clicked',
    PROJECTS_NAV_CLICKED: 'projects_nav_clicked',
    EXPERIENCE_NAV_CLICKED: 'experience_nav_clicked',
    THEME_TOGGLE_CLICKED: 'theme_toggle_clicked',
    RESUME_CLICKED: 'resume_clicked',
    CONTACT_ME_CLICKED: 'contact_me_clicked',
    GITHUB_CLICKED: 'github_icon_clicked',
    LINKEDIN_CLICKED: 'linkedin_icon_clicked',
    SPOTIFY_CLICKED: 'spotify_icon_clicked',
    MEDIUM_CLICKED: 'medium_icon_clicked',
    MAIL_CLICKED: 'mail_icon_clicked',
};

export const navItems = [
    {
        to: 'about',
        num: '01.',
        name: 'About',
        clickEventId: gaEvents.ABOUT_NAV_CLICKED,
    },
    {
        to: 'projects',
        num: '02.',
        name: 'Projects',
        clickEventId: gaEvents.PROJECTS_NAV_CLICKED,
    },
    {
        to: 'experience',
        num: '03.',
        name: 'Experience',
        clickEventId: gaEvents.EXPERIENCE_NAV_CLICKED,
    },
];

export const projects = [
    {
        nameId: 'PROJECTS_PATCH',
        descriptionId: 'PROJECTS_PATCH_DESCRIPTION',
        githubLink: 'https://github.com/willshiao/Patch',
    },
    {
        nameId: 'PROJECTS_CHEATERBEATER',
        descriptionId: 'PROJECTS_CHEATERBEATER_DESCRIPTION',
        githubLink: 'https://github.com/willshiao/CheaterBeater',
    },
    {
        nameId: 'PROJECTS_BAUT',
        descriptionId: 'PROJECTS_BAUT_DESCRIPTION',
        githubLink: 'https://github.com/KirbyDownB/GEICOChatBot',
    },
    {
        nameId: 'PROJECTS_NUTSHELL',
        descriptionId: 'PROJECTS_NUTSHELL_DESCRIPTION',
        githubLink: 'https://github.com/chefong/nutshell',
    },
    {
        nameId: 'PROJECTS_STOCK_OVERFLOW',
        descriptionId: 'PROJECTS_STOCK_OVERFLOW_DESCRIPTION',
        githubLink: 'https://github.com/kimjihwan0208/stockoverflow',
    },
    {
        nameId: 'PROJECTS_JOYRIDE',
        descriptionId: 'PROJECTS_JOYRIDE_DESCRIPTION',
        githubLink: 'https://github.com/ericong18/joyride',
    },
];

export const experiences = [
    {
        imagePath: {
            light: 'imgs/experience/opentable.png',
            dark: 'imgs/experience/opentable-white.png',
        },
        nameId: 'EXPERIENCE_OPENTABLE',
        positionId: 'SOFTWARE_ENGINEER_1',
        duration: 'February 2021 - Present',
        teamId: 'EXPERIENCE_OPENTABLE_TEAM_2',
        initiativesId: 'EXPERIENCE_OPENTABLE_INITIATIVES_2',
        technologiesId: 'EXPERIENCE_OPENTABLE_TECHNOLOGIES_2',
    },
    {
        imagePath: {
            light: '/imgs/experience/intuit.svg',
            dark: '/imgs/experience/intuit-white.svg',
        },
        nameId: 'EXPERIENCE_INTUIT',
        positionId: 'SOFTWARE_ENGINEER_1',
        duration: 'July 2020 - February 2021',
        teamId: 'EXPERIENCE_INTUIT_TEAM',
        initiativesId: 'EXPERIENCE_INTUIT_INITIATIVES',
        technologiesId: 'EXPERIENCE_INTUIT_TECHNOLOGIES',
    },
    {
        imagePath: {
            light: 'imgs/experience/opentable.png',
            dark: 'imgs/experience/opentable-white.png',
        },
        nameId: 'EXPERIENCE_OPENTABLE',
        positionId: 'SOFTWARE_ENGINEERING_INTERN',
        duration: 'June 2019 - August 2019',
        teamId: 'EXPERIENCE_OPENTABLE_TEAM_1',
        initiativesId: 'EXPERIENCE_OPENTABLE_INITIATIVES_1',
        technologiesId: 'EXPERIENCE_OPENTABLE_TECHNOLOGIES_1',
    },
];

export const handles = [
    {
        name: 'GitHub',
        icon: ['fab', 'github'],
        link: 'https://github.com/chefong',
        clickEventId: gaEvents.GITHUB_CLICKED,
    },
    {
        name: 'LinkedIn',
        icon: ['fab', 'linkedin-in'],
        link: 'https://www.linkedin.com/in/ericong18/',
        clickEventId: gaEvents.LINKEDIN_CLICKED,
    },
    {
        name: 'Spotify',
        icon: ['fab', 'spotify'],
        link: 'https://open.spotify.com/user/ericong18?si=H0kHkFdxQ-GyKIlfLWzDqg',
        clickEventId: gaEvents.SPOTIFY_CLICKED,
    },
    {
        name: 'Medium',
        icon: ['fab', 'medium'],
        link: 'https://medium.com/@ericong18',
        clickEventId: gaEvents.MEDIUM_CLICKED,
    },
    {
        name: 'Email',
        icon: 'envelope',
        link: 'mailto:ericong18@gmail.com',
        clickEventId: gaEvents.MAIL_CLICKED,
    },
];

export const likes = [
    'ABOUT_LIKE_TEA',
    'ABOUT_LIKE_WATERMELON',
    'ABOUT_LIKE_MUSIC',
    'ABOUT_LIKE_DESIGN',
    'ABOUT_LIKE_PREMIER_LEAGUE',
];

export const dislikes = ['ABOUT_DISLIKE_CILANTRO'];

export const themeColors = {
    light: '#fffefb',
    dark: '#222222',
    tan: '#f1e7d0',
};

export const lottieProps = {
    loop: false,
    autoplay: false,
    animationData: hamburgerMenuAnimation,
};

export const reactScrollLinkProps = {
    smooth: true,
    offset: -150,
    duration: 750,
};

// Segment frames for menu open and close
export const menuOpenSegments = [30, 70];
export const menuCloseSegments = [100, 140];

// Box shadow styling for when the nav menu is open
export const menuBoxShadow = '0px 4px 28px rgba(0, 0, 0, 0.06)';

// Menu transition styling for menu
export const menuTransition = 'all 0.5s ease';

export const themeOrder = ['light', 'tan', 'dark'];

export const iconSize = '28px';

export const iconColors = {
    light: themeColors.dark,
    dark: themeColors.light,
    tan: themeColors.dark,
};

export const toastOptions = {
    style: {
        padding: '16px',
        fontFamily: 'Greycliff',
        backgroundColor: '#303030',
        color: 'white',
    },
};

export const particlesParams = {
    particles: {
        number: {
            value: 80,
            density: {
                enable: true,
                value_area: 1500,
            },
        },
        line_linked: {
            enable: true,
            opacity: 0.02,
        },
        move: {
            direction: 'right',
            speed: 0.05,
        },
        size: {
            value: 1.25,
        },
        opacity: {
            anim: {
                enable: true,
                speed: 1,
                opacity_min: 0.05,
            },
        },
    },
    retina_detect: true,
};
