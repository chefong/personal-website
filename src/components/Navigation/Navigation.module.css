.logo {
  width: 85%;
}

.navbar, .navCollapse {
  background-color: #fffefb;
}

.navLink {
  font-weight: 500;
  font-size: 1.2em;
  color: #303030 !important;
} 

.navLink:hover {
  cursor: pointer;
}

.navNum {
  color: #4A74E1;
}

.toggler {
  border: none;
}

.menuIcon {
  width: 48px;
}