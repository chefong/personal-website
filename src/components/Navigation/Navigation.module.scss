.logo {
  width: 85%;
}

.navbar, .navCollapse {
  background-color: var(--theme-page-background);
}

.navLink {
  font-weight: bold;
  font-size: 1.2em;
  color: var(--theme-page-text) !important;
} 

.navLink:hover {
  cursor: pointer;
}

.navNum {
  color: var(--theme-color);
}

.toggler {
  border: none;
}

.menuIcon {
  width: 48px;
}

.toggle {
  margin-left: 24px;
  margin-top: 6px;
  margin-bottom: 10px;
}