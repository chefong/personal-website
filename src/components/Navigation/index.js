import { useState, useEffect, useRef } from 'react';
import { Link } from 'react-scroll';
import Lottie from 'lottie-react';
import {
  Collapse,
  Navbar,
  NavbarToggler,
  NavbarBrand,
  Nav,
  NavItem,
} from 'reactstrap';
import {
  navItems,
  lottieProps,
  reactScrollLinkProps,
  menuOpenSegments,
  menuCloseSegments,
  menuBoxShadow,
  menuTransition,
} from '~/common/constants';
import styles from './styles.module.scss';
import ThemeToggle from '~/components/ThemeToggle';
import { logToGA } from '~/common/utils/ga';

export default function Navigation() {
  const [isOpen, setIsOpen] = useState(false);
  const [scrollPosition, setScrollPosition] = useState(0);
  const menuIconEl = useRef(null);

  const openMenu = () => {
    menuIconEl.current.playSegments(menuOpenSegments, true);
    setIsOpen(true);
  };

  const closeMenu = () => {
    menuIconEl.current.playSegments(menuCloseSegments, true);
    setIsOpen(false);
  };

  const toggle = () => (isOpen ? closeMenu() : openMenu());

  const logNavItemClick = (clickEventId) => {
    logToGA({ action: clickEventId });
  };

  useEffect(() => window.addEventListener('scroll', () => setScrollPosition(window.scrollY)), []);

  return (
    <div data-testid="Navigation-container">
      <Navbar
        light
        expand="md"
        fixed="top"
        className={styles.navbar}
        style={{
          boxShadow: (scrollPosition > 0 || isOpen) ? menuBoxShadow : 'none',
          transition: menuTransition,
        }}
      >
        <NavbarBrand href="/" className="ml-2">
          <img className={styles.logo} src="/imgs/logo.svg" alt="" />
        </NavbarBrand>
        <NavbarToggler onClick={toggle} className={styles.toggler} data-testid="Navigation-toggle">
          <Lottie
            {...lottieProps}
            lottieRef={menuIconEl}
            className={styles.menuIcon}
          />
        </NavbarToggler>
        <Collapse isOpen={isOpen} navbar data-testid="Navigation-collapse">
          <Nav className="ml-auto mr-5 justify-content-center align-items-md-center" navbar>
            {navItems.map(({
              to, num, name, clickEventId,
            }) => (
              <NavItem className="ml-sm-3 ml-md-5 mt-2 mb-2" key={name} data-testid="Navigation-item">
                <Link
                  {...reactScrollLinkProps}
                  className={styles.navLink}
                  to={to}
                  onClick={() => logNavItemClick(clickEventId)}
                >
                  <span className={styles.navNum}>{num}</span>
                  {' '}
                  {name}
                </Link>
              </NavItem>
            ))}
            <NavItem className={`${styles.toggle} ml-sm-3 ml-md-5`}>
              <ThemeToggle />
            </NavItem>
          </Nav>
        </Collapse>
      </Navbar>
    </div>
  );
}
