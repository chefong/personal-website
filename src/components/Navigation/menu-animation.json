{"v":"5.1.8","fr":59.9400024414063,"ip":0,"op":170.000006924242,"w":512,"h":512,"nm":"Icon","ddd":0,"assets":[],"layers":[{"ddd":0,"ind":1,"ty":4,"nm":"Line 3","sr":1,"ks":{"o":{"a":0,"k":100,"ix":11},"r":{"a":1,"k":[{"i":{"x":[0.2],"y":[1]},"o":{"x":[0.8],"y":[0]},"n":["0p2_1_0p8_0"],"t":50,"s":[0],"e":[45]},{"i":{"x":[0.2],"y":[1]},"o":{"x":[0.8],"y":[0]},"n":["0p2_1_0p8_0"],"t":70,"s":[45],"e":[45]},{"i":{"x":[0.2],"y":[1]},"o":{"x":[0.8],"y":[0]},"n":["0p2_1_0p8_0"],"t":100,"s":[45],"e":[0]},{"t":120.0000048877}],"ix":10},"p":{"a":1,"k":[{"i":{"x":0.2,"y":1},"o":{"x":0.8,"y":0},"n":"0p2_1_0p8_0","t":30,"s":[256,149.333,0],"e":[256,256,0],"to":[0,17.7777500152588,0],"ti":[0,-17.7777500152588,0]},{"i":{"x":0.2,"y":0.2},"o":{"x":0.8,"y":0.8},"n":"0p2_0p2_0p8_0p8","t":50,"s":[256,256,0],"e":[256,256,0],"to":[0,0,0],"ti":[0,0,0]},{"i":{"x":0.2,"y":1},"o":{"x":0.8,"y":0},"n":"0p2_1_0p8_0","t":120,"s":[256,256,0],"e":[256,149.333,0],"to":[0,-17.7777500152588,0],"ti":[0,17.7777500152588,0]},{"t":140.000005702317}],"ix":2},"a":{"a":0,"k":[192.25,21.583,0],"ix":1},"s":{"a":0,"k":[100,100,100],"ix":6}},"ao":0,"shapes":[{"ty":"gr","it":[{"ind":0,"ty":"sh","ix":1,"ks":{"a":0,"k":{"i":[[0,0],[0,0],[0,0],[0,0]],"o":[[0,0],[0,0],[0,0],[0,0]],"v":[[-192,21.333],[192,21.333],[192,-21.333],[-192,-21.333]],"c":true},"ix":2},"nm":"Path 1","mn":"ADBE Vector Shape - Group","hd":false},{"ty":"fl","c":{"a":0,"k":[0.258999992819,0.258999992819,0.258999992819,1],"ix":4},"o":{"a":0,"k":100,"ix":5},"r":1,"nm":"Fill 1","mn":"ADBE Vector Graphic - Fill","hd":false},{"ty":"tr","p":{"a":0,"k":[192.25,21.583],"ix":2},"a":{"a":0,"k":[0,0],"ix":1},"s":{"a":0,"k":[100,100],"ix":3},"r":{"a":0,"k":0,"ix":6},"o":{"a":0,"k":100,"ix":7},"sk":{"a":0,"k":0,"ix":4},"sa":{"a":0,"k":0,"ix":5},"nm":"Transform"}],"nm":"Group 1","np":2,"cix":2,"ix":1,"mn":"ADBE Vector Group","hd":false}],"ip":0,"op":170.000006924242,"st":0,"bm":0},{"ddd":0,"ind":2,"ty":4,"nm":"Line 4","sr":1,"ks":{"o":{"a":0,"k":100,"ix":11},"r":{"a":0,"k":0,"ix":10},"p":{"a":0,"k":[256,255.999,0],"ix":2},"a":{"a":0,"k":[192.25,21.584,0],"ix":1},"s":{"a":0,"k":[100,100,100],"ix":6}},"ao":0,"shapes":[{"ty":"gr","it":[{"ind":0,"ty":"sh","ix":1,"ks":{"a":0,"k":{"i":[[0,0],[0,0],[0,0],[0,0]],"o":[[0,0],[0,0],[0,0],[0,0]],"v":[[-192,21.334],[192,21.334],[192,-21.334],[-192,-21.334]],"c":true},"ix":2},"nm":"Path 1","mn":"ADBE Vector Shape - Group","hd":false},{"ty":"fl","c":{"a":0,"k":[0.258999992819,0.258999992819,0.258999992819,1],"ix":4},"o":{"a":0,"k":100,"ix":5},"r":1,"nm":"Fill 1","mn":"ADBE Vector Graphic - Fill","hd":false},{"ty":"tr","p":{"a":0,"k":[192.25,21.584],"ix":2},"a":{"a":0,"k":[0,0],"ix":1},"s":{"a":0,"k":[100,100],"ix":3},"r":{"a":0,"k":0,"ix":6},"o":{"a":0,"k":100,"ix":7},"sk":{"a":0,"k":0,"ix":4},"sa":{"a":0,"k":0,"ix":5},"nm":"Transform"}],"nm":"Group 1","np":2,"cix":2,"ix":1,"mn":"ADBE Vector Group","hd":false}],"ip":120.0000048877,"op":170.000006924242,"st":120.0000048877,"bm":0},{"ddd":0,"ind":3,"ty":4,"nm":"Line 2","sr":1,"ks":{"o":{"a":0,"k":100,"ix":11},"r":{"a":0,"k":0,"ix":10},"p":{"a":0,"k":[256,255.999,0],"ix":2},"a":{"a":0,"k":[192.25,21.584,0],"ix":1},"s":{"a":0,"k":[100,100,100],"ix":6}},"ao":0,"shapes":[{"ty":"gr","it":[{"ind":0,"ty":"sh","ix":1,"ks":{"a":0,"k":{"i":[[0,0],[0,0],[0,0],[0,0]],"o":[[0,0],[0,0],[0,0],[0,0]],"v":[[-192,21.334],[192,21.334],[192,-21.334],[-192,-21.334]],"c":true},"ix":2},"nm":"Path 1","mn":"ADBE Vector Shape - Group","hd":false},{"ty":"fl","c":{"a":0,"k":[0.258999992819,0.258999992819,0.258999992819,1],"ix":4},"o":{"a":0,"k":100,"ix":5},"r":1,"nm":"Fill 1","mn":"ADBE Vector Graphic - Fill","hd":false},{"ty":"tr","p":{"a":0,"k":[192.25,21.584],"ix":2},"a":{"a":0,"k":[0,0],"ix":1},"s":{"a":0,"k":[100,100],"ix":3},"r":{"a":0,"k":0,"ix":6},"o":{"a":0,"k":100,"ix":7},"sk":{"a":0,"k":0,"ix":4},"sa":{"a":0,"k":0,"ix":5},"nm":"Transform"}],"nm":"Group 1","np":2,"cix":2,"ix":1,"mn":"ADBE Vector Group","hd":false}],"ip":0,"op":51.0000020772726,"st":0,"bm":0},{"ddd":0,"ind":4,"ty":4,"nm":"Line 1","sr":1,"ks":{"o":{"a":0,"k":100,"ix":11},"r":{"a":1,"k":[{"i":{"x":[0.2],"y":[1]},"o":{"x":[0.8],"y":[0]},"n":["0p2_1_0p8_0"],"t":50,"s":[0],"e":[-45]},{"i":{"x":[0.2],"y":[1]},"o":{"x":[0.8],"y":[0]},"n":["0p2_1_0p8_0"],"t":70,"s":[-45],"e":[-45]},{"i":{"x":[0.2],"y":[1]},"o":{"x":[0.8],"y":[0]},"n":["0p2_1_0p8_0"],"t":100,"s":[-45],"e":[0]},{"t":120.0000048877}],"ix":10},"p":{"a":1,"k":[{"i":{"x":0.2,"y":1},"o":{"x":0.8,"y":0},"n":"0p2_1_0p8_0","t":30,"s":[256,359.667,0],"e":[256,256,0],"to":[0,-17.2777500152588,0],"ti":[0,17.2777500152588,0]},{"i":{"x":0.2,"y":0.2},"o":{"x":0.8,"y":0.8},"n":"0p2_0p2_0p8_0p8","t":50,"s":[256,256,0],"e":[256,256,0],"to":[0,0,0],"ti":[0,0,0]},{"i":{"x":0.2,"y":1},"o":{"x":0.8,"y":0},"n":"0p2_1_0p8_0","t":120,"s":[256,256,0],"e":[256,359.667,0],"to":[0,17.2777500152588,0],"ti":[0,-17.2777500152588,0]},{"t":140.000005702317}],"ix":2},"a":{"a":0,"k":[192.25,21.583,0],"ix":1},"s":{"a":0,"k":[100,100,100],"ix":6}},"ao":0,"shapes":[{"ty":"gr","it":[{"ind":0,"ty":"sh","ix":1,"ks":{"a":0,"k":{"i":[[0,0],[0,0],[0,0],[0,0]],"o":[[0,0],[0,0],[0,0],[0,0]],"v":[[-192,21.333],[192,21.333],[192,-21.333],[-192,-21.333]],"c":true},"ix":2},"nm":"Path 1","mn":"ADBE Vector Shape - Group","hd":false},{"ty":"fl","c":{"a":0,"k":[0.258999992819,0.258999992819,0.258999992819,1],"ix":4},"o":{"a":0,"k":100,"ix":5},"r":1,"nm":"Fill 1","mn":"ADBE Vector Graphic - Fill","hd":false},{"ty":"tr","p":{"a":0,"k":[192.25,21.583],"ix":2},"a":{"a":0,"k":[0,0],"ix":1},"s":{"a":0,"k":[100,100],"ix":3},"r":{"a":0,"k":0,"ix":6},"o":{"a":0,"k":100,"ix":7},"sk":{"a":0,"k":0,"ix":4},"sa":{"a":0,"k":0,"ix":5},"nm":"Transform"}],"nm":"Group 1","np":2,"cix":2,"ix":1,"mn":"ADBE Vector Group","hd":false}],"ip":0,"op":170.000006924242,"st":0,"bm":0}],"markers":[]}