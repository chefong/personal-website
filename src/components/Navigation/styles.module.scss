.logo {
  width: 48px;
}

.navbar, .navCollapse {
  background-color: var(--theme-page-background);
}

.navbarContainer {
  width: 100%;
  max-width: 1440px;
  display: flex;
  margin: 0 auto;
  padding: 0 16px;
}

.navbar {
  padding: 16px 32px;
  box-sizing: border-box;
}

.navbarBrand {
  margin: 0;
}

.nav {
  flex: 1;
  display: flex;
  align-items: center;
  justify-content: flex-end;
  gap: 32px;
}

.navLink {
  font-weight: 800;
  font-size: 1.2em;
  color: var(--theme-page-text) !important;

  &:hover {
    text-decoration: underline;
    cursor: pointer;
  }

  @media (max-width: 576px) {
    margin-left: 16px;
  }
} 

.navLink:hover {
  cursor: pointer;
}

.navNum {
  color: var(--theme-color);
}

.toggler {
  border: none;
}

.menuIcon {
  width: 48px;
}

.toggle {
  @media (max-width: 576px) {
    margin-top: 4px;
    margin-bottom: 8px;
    margin-left: 1rem !important;
  }
}
