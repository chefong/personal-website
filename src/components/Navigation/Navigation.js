import { useState, useEffect, useRef } from 'react';
import { Link } from 'react-scroll';
/* istanbul ignore next */
import Lottie from 'lottie-react';
import { navItems } from '../../shared/constants';
import {
  Collapse,
  Navbar,
  NavbarToggler,
  NavbarBrand,
  Nav,
  NavItem
} from 'reactstrap';
import styles from './Navigation.module.css';
import menuAnimation from './menu-animation.json';

const lottieProps = {
  loop: false,
  autoplay: false,
  animationData: menuAnimation
};

// Segment frames for menu open and close
const menuOpenSegments = [30, 70];
const menuCloseSegments = [100, 140];

export default function Navigation() {
  const [isOpen, setIsOpen] = useState(false);
  const [scrollPosition, setScrollPosition] = useState(0);
  const menuIconEl = useRef(null);

  const toggle = () => {
    if (isOpen) {
      menuIconEl.current.playSegments(menuCloseSegments, true);
      setIsOpen(false);
    } else {
      menuIconEl.current.playSegments(menuOpenSegments, true);
      setIsOpen(true);
    }
  };

  /* istanbul ignore next */
  useEffect(() => window.addEventListener('scroll', () => setScrollPosition(window.scrollY)), []);

  return (
    <div data-testid="Navigation-container">
      <Navbar
        light
        expand="md"
        fixed="top"
        className={styles.navbar}
        style={{
          boxShadow: (scrollPosition > 0 || isOpen) ? "0px 4px 28px rgba(0, 0, 0, 0.06)" : "none",
          transition: "all 0.5s ease"
        }}
      >
        <NavbarBrand href="/" className="ml-2">
          <img className={styles.logo} src="/imgs/logo.svg" alt=""/>
        </NavbarBrand>
        <NavbarToggler onClick={toggle} className={styles.toggler} data-testid="Navigation-toggle">
          <Lottie
            {...lottieProps}
            lottieRef={menuIconEl}
            className={styles.menuIcon}
          />
        </NavbarToggler>
        <Collapse isOpen={isOpen} navbar data-testid="Navigation-collapse">
          <Nav className="ml-auto mr-5 mt-3 mb-3" navbar>
            {navItems.map(({ to, num, name }, index) => (
              <NavItem className="ml-4 mt-2 mb-2" key={index} data-testid="Navigation-item">
                <Link smooth className={styles.navLink} to={to} offset={-150} duration={750}>
                  <span className={styles.navNum}>{num}</span> {name}
                </Link>
              </NavItem>
            ))}
          </Nav>
        </Collapse>
      </Navbar>
    </div>
  );
}