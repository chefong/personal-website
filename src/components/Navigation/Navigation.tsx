import { useState, useEffect, useRef, ReactElement } from 'react';
import { Link } from 'react-scroll';
import Lottie from 'lottie-react';
import { navItems } from '../../shared/constants';
import {
  Collapse,
  Navbar,
  NavbarToggler,
  NavbarBrand,
  Nav,
  NavItem
} from 'reactstrap';
import styles from './Navigation.module.css';
import menuAnimation from './menu-animation.json';
import { LottieSegment } from '../../interfaces';

const lottieProps = {
  loop: false,
  autoplay: false,
  animationData: menuAnimation
};

// Segment frames for menu open and close
const menuOpenSegments: LottieSegment = { start: 30, end: 70};
const menuCloseSegments: LottieSegment = { start: 100, end: 140 };

export default function Navigation(): ReactElement {
  const [isOpen, setIsOpen] = useState(false);
  const [scrollPosition, setScrollPosition] = useState(0);
  const menuIconEl = useRef(null);

  const toggle = () => {
    if (isOpen) {
      const { start, end } = menuCloseSegments;
      menuIconEl.current.playSegments([start, end], true);
      setIsOpen(false);
    } else {
      const { start, end } = menuOpenSegments;
      menuIconEl.current.playSegments([start, end], true);
      setIsOpen(true);
    }
  };

  /* istanbul ignore next */
  useEffect(() => window.addEventListener('scroll', () => setScrollPosition(window.scrollY)), []);

  return (
    <div data-testid="Navigation-container">
      <Navbar
        light
        expand="md"
        fixed="top"
        className={styles.navbar}
        style={{
          boxShadow: (scrollPosition > 0 || isOpen) ? "0px 4px 28px rgba(0, 0, 0, 0.06)" : "none",
          transition: "all 0.5s ease"
        }}
      >
        <NavbarBrand href="/" className="ml-2">
          <img className={styles.logo} src="/imgs/logo.svg" alt=""/>
        </NavbarBrand>
        <NavbarToggler onClick={toggle} className={styles.toggler} data-testid="Navigation-toggle">
          <Lottie
            {...lottieProps}
            lottieRef={menuIconEl}
            className={styles.menuIcon}
          />
        </NavbarToggler>
        <Collapse isOpen={isOpen} navbar data-testid="Navigation-collapse">
          <Nav className="ml-auto mr-5 mt-3 mb-3" navbar>
            {navItems.map(({ to, num, name }, index) => (
              <NavItem className="ml-4 mt-2 mb-2" key={index} data-testid="Navigation-item">
                <Link smooth className={styles.navLink} to={to} offset={-150} duration={750}>
                  <span className={styles.navNum}>{num}</span> {name}
                </Link>
              </NavItem>
            ))}
          </Nav>
        </Collapse>
      </Navbar>
    </div>
  );
}