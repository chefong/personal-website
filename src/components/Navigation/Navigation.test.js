import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import Navigation from './Navigation';
import { navItems } from '../../shared/constants';

describe('Navigation', () => {
  it('renders', () => {
    const { getByTestId } = render(<Navigation />);
    expect(getByTestId('Navigation-container')).toBeDefined();
  });

  it('matches snapshot', () => {
    const { container } = render(<Navigation />);
    expect(container).toMatchSnapshot();
  });

  it('contains the correct number of navigation items', () => {
    const { getAllByTestId } = render(<Navigation />);
    expect(getAllByTestId('Navigation-item')).toHaveLength(navItems.length);
  });

  it('can toggle menu navigation', () => {
    const { getByTestId } = render(<Navigation />);
    
    fireEvent.click(getByTestId('Navigation-toggle'));
    expect(getByTestId('Navigation-collapse')).toHaveClass('collapsing');

    fireEvent.click(getByTestId('Navigation-toggle'));
    waitFor(() => 
      {
        debug();
        expect(getByTestId('Navigation-collapse')).not.toHaveClass('collapsing');
      },
      {
        timeout: 1000
      }
    );
  })
});