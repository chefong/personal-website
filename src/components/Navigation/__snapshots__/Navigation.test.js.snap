// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation matches snapshot 1`] = `
<div>
  <div
    data-testid="Navigation-container"
  >
    <nav
      class="navbar navbar-expand-md navbar-light fixed-top"
      style="box-shadow: none; transition: all 0.5s ease;"
    >
      <a
        class="ml-2 navbar-brand"
        href="/"
      >
        <img
          alt=""
          src="/imgs/logo.svg"
        />
      </a>
      <button
        aria-label="Toggle navigation"
        class="navbar-toggler"
        data-testid="Navigation-toggle"
        type="button"
      >
        <div>
          <svg
            height="512"
            preserveAspectRatio="xMidYMid meet"
            style="width: 100%; height: 100%; transform: translate3d(0,0,0);"
            viewBox="0 0 512 512"
            width="512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <defs>
              <clippath
                id="__lottie_element_16"
              >
                <rect
                  height="512"
                  width="512"
                  x="0"
                  y="0"
                />
              </clippath>
            </defs>
            <g
              clip-path="url(#__lottie_element_16)"
            >
              <g
                opacity="1"
                style="display: block;"
                transform="matrix(1,0,0,1,63.75,338.083984375)"
              >
                <g
                  opacity="1"
                  transform="matrix(1,0,0,1,192.25,21.58300018310547)"
                >
                  <path
                    d=" M-192,21.33300018310547 C-192,21.33300018310547 192,21.33300018310547 192,21.33300018310547 C192,21.33300018310547 192,-21.33300018310547 192,-21.33300018310547 C192,-21.33300018310547 -192,-21.33300018310547 -192,-21.33300018310547 C-192,-21.33300018310547 -192,21.33300018310547 -192,21.33300018310547z"
                    fill="rgb(66,66,66)"
                    fill-opacity="1"
                  />
                </g>
              </g>
              <g
                opacity="1"
                style="display: block;"
                transform="matrix(1,0,0,1,63.75,234.4149932861328)"
              >
                <g
                  opacity="1"
                  transform="matrix(1,0,0,1,192.25,21.583999633789062)"
                >
                  <path
                    d=" M-192,21.333999633789062 C-192,21.333999633789062 192,21.333999633789062 192,21.333999633789062 C192,21.333999633789062 192,-21.333999633789062 192,-21.333999633789062 C192,-21.333999633789062 -192,-21.333999633789062 -192,-21.333999633789062 C-192,-21.333999633789062 -192,21.333999633789062 -192,21.333999633789062z"
                    fill="rgb(66,66,66)"
                    fill-opacity="1"
                  />
                </g>
              </g>
              <g
                style="display: none;"
              >
                <g>
                  <path />
                </g>
              </g>
              <g
                opacity="1"
                style="display: block;"
                transform="matrix(1,0,0,1,63.75,127.74999237060547)"
              >
                <g
                  opacity="1"
                  transform="matrix(1,0,0,1,192.25,21.58300018310547)"
                >
                  <path
                    d=" M-192,21.33300018310547 C-192,21.33300018310547 192,21.33300018310547 192,21.33300018310547 C192,21.33300018310547 192,-21.33300018310547 192,-21.33300018310547 C192,-21.33300018310547 -192,-21.33300018310547 -192,-21.33300018310547 C-192,-21.33300018310547 -192,21.33300018310547 -192,21.33300018310547z"
                    fill="rgb(66,66,66)"
                    fill-opacity="1"
                  />
                </g>
              </g>
            </g>
          </svg>
        </div>
      </button>
      <div
        class="collapse navbar-collapse"
        data-testid="Navigation-collapse"
      >
        <ul
          class="ml-auto mr-5 mt-3 mb-3 navbar-nav"
        >
          <li
            class="ml-4 mt-2 mb-2 nav-item"
            data-testid="Navigation-item"
          >
            <a>
              <span>
                01.
              </span>
               
              About
            </a>
          </li>
          <li
            class="ml-4 mt-2 mb-2 nav-item"
            data-testid="Navigation-item"
          >
            <a>
              <span>
                02.
              </span>
               
              Projects
            </a>
          </li>
          <li
            class="ml-4 mt-2 mb-2 nav-item"
            data-testid="Navigation-item"
          >
            <a>
              <span>
                03.
              </span>
               
              Experience
            </a>
          </li>
        </ul>
      </div>
    </nav>
  </div>
</div>
`;
